src/python/WMCore/Storage/Backends/GFAL2Impl.py

```diff
@@ -113,27 +109,89 @@ def createStageOutCommand(self, sourcePFN, targetPFN, options=None, checksums=No
         copyCommandDict['destination'] = self.createFinalPFN(targetPFN)
 
+        return copyCommandDict
+
+    def createStageOutCommand(self, sourcePFN, targetPFN, options=None, checksums=None):
+        """
+        Create gfal-cp command for stageOut
+
+        :sourcePFN: str, PFN of the source file
+        :targetPFN: str, destination PFN
+        :options: str, additional options for gfal-cp
+        :checksums: dict, collect checksums according to the algorithms saved as keys
+        """
+
+
+        copyCommandDict = self.buildCopyCommandDict(sourcePFN, targetPFN, options, checksums)
         copyCommand = self.copyCommand % copyCommandDict
-        result += copyCommand
+        result = "#!/bin/bash\n" + copyCommand
 
         if _CheckExitCodeOption:
             result += """
             EXIT_STATUS=$?
             echo "gfal-copy exit status: $EXIT_STATUS"
             if [[ $EXIT_STATUS != 0 ]]; then
-               echo "ERROR: gfal-copy exited with $EXIT_STATUS"
-               echo "Cleaning up failed file:"
-               %s
+                echo "ERROR: gfal-copy exited with $EXIT_STATUS"
+                echo "Cleaning up failed file:"
+                {remove_command}
             fi
             exit $EXIT_STATUS
-            """ % self.createRemoveFileCommand(targetPFN)
-
+            """.format(
+                remove_command=self.createRemoveFileCommand(targetPFN)
+            )
+        
         return result
 
+    def createDebuggingCommand(self, sourcePFN, targetPFN, options=None, checksums=None):
+        """
+        Debug a failed gfal-cp command for stageOut, without re-running it,
+        providing information on the environment and the certifications
+
+        :sourcePFN: str, PFN of the source file
+        :targetPFN: str, destination PFN
+        :options: str, additional options for gfal-cp
+        :checksums: dict, collect checksums according to the algorithms saved as keys
+        """
+        
+        copyCommandDict = self.buildCopyCommandDict(sourcePFN, targetPFN, options, checksums)
+        copyCommand = self.copyCommand % copyCommandDict
+
+        try:
+            stdout, stderr, returnCode = runCommand("voms-proxy-info -file $X509_USER_PROXY", timeout=10)
+            if returnCode == 0:
+                logging.debug("Success! voms-proxy-info output:\n%s", stdout)       
+            else:
+                logging.debug("Failure! voms-proxy-info failed with return code: %s", returnCode)
+                logging.debug("Error output: %s", stderr)
+        except SubprocessAlgoException as e:
+            logging.debug("An exception occurred while running voms-proxy-info:")
+            logging.debug(e)
+
+        logging.debug("Actual gfal-cp command which failed: %s", copyCommand)
+
+        result = "#!/bin/bash\n"
+        result += """
+        echo "gfal-cp command which failed: {copy_command}"
+        echo "gfal-cp location:"
```

Comment:
Let me suggest a minor enhancement here: `echo "gfal-cp location: $(which gfal-cp)"`.
Similarly to the version command below.